src/efftester.ml

```diff
@@ -1210,15 +1381,16 @@ module Shrinker = struct
     | _ -> Iter.empty
   ;;
 
-  let ( <+> ) = Iter.( <+> )
-
   let rec term_shrinker term =
+    let ( <+> ) = Iter.( <+> ) in
     match term with
     | Lit l -> shrink_lit l
     | Variable (t, _) ->
       (match create_lit t with
       | Some c -> Iter.return c
       | _ -> Iter.empty)
+    | Constructor _ as c -> Iter.return c
```

Comment:
In the case `Some e`, you could (1 shrink this into `None`, or you could (2) shrink `e` recursively.

Notice that (2) is a uniform idea that applies to all `Constructor` terms (no matter at which type they are), while (1) is specific to `option` and it depends of what we know of the constructors of option type (we know that `None` and `Some e` return the same type, and that `None` is "smaller" than `Some e`).

I would recommend writing a generic shrink that just shrinks the parameters recursively (note: it cannot shrink constructors without parameters), and taking the union (`<+>`) with a constructor-specific shrinker that matches on the constructor name and does something only on the builtin constructors (for the others, it just fails to shrink). This structure will make it easier to extend the code later.

Reply:
Here would be an attempt (I didn't try it or even type-check) at this logic:

```ocaml
| Constructor (ty, cstr, args) ->
  let generic_shrink =
    Iter.map (fun shrunk_args -> Constructor(ty, cstr, shrunk_args)) (Shrink.list args) in
  let specific_thrink =
    match cstr with
    | "None" -> Iter.empty
    | "Some" -> Iter.return (Constructor (ty, "None", []))
    | _ -> (* unknown constructor *) Iter.empty
  in
  generic_shrink <+> specific_shrink
```

Reply:
Note: `Shrink.list` tries to reduce the number of elements in the list (which in our case would introduce a type-checking error!) before getting to reducing the individual list elements. Here's a size-preserving shrinker (from other work) that could be useful instead:
```ocaml
(* a size-preserving list-shrinker *)
let rec shrink_list_elements ~shrink xs = match xs with
  | [] -> Iter.empty
  | x::xs ->
    Iter.(
      map (fun x' -> x'::xs) (shrink x)
      <+>
      map (fun xs -> x::xs) (shrink_list_elements ~shrink xs))
```
I guess this could also be a useful addition to QCheck... 😀

Reply:
Actually: Calling `Shrink.list` without the optional element shrinker `~shrink` argument can only reduce the list size - which is not what we want, IMO.

Reply:
(I noticed the issue independently from @jmid above, so my comment below is written without taking his comment into account.)

Actually 

```ocaml
  let generic_shrink =
    Iter.map (fun shrunk_args -> Constructor(ty, cstr, shrunk_args)) (Shrink.list args) in
```

is not correct (it will remove arguments instead of recursively shrinking them), we miss a traverse/sequence combinator. I just sent a Pull Request to QCheck to implement it: https://github.com/c-cube/qcheck/pull/68 . (We could implement it internally in our codebase but this is of interest to any QCheck user.)

Reply:
@jmid: in the PR I sent, all combinations of possible shrinking choices are returned. Your approach, shrinking only one element of the list each time, may be much more efficient. I would encourage you to comment on alternative implementations on the QCheck PR that I sent.

Reply:
Thanks a lot @gasche for your PR because it was causing a quite perplexing problem for me to get shrink a list and get `'a list Iter.t`.